### PR TITLE
allow empty filter values in report builder reports

### PR DESCRIPTION
**Pull request by czue**
https://trello.com/c/xRBl3440/46-report-builder-upgrades / https://docs.google.com/document/d/1cRFJ3AKGHnJRO2sLaTEOW4o9Z-jOYejONzv_jjW6X8Y/edit#

##### SUMMARY

This supports specifying "empty" values in dynamic choice list filters (affecting report builder reports and UCRs), allowing users to filter data where the property is is empty or not set.

The filters look like this:
image

And only render if there is underlying data that is either `null` or an empty string, respectively.

I'm proposing releasing this to everyone, though could put behind a flag if that doesn't seem right.

I didn't actually dig into what causes a UCR to have null data versus empty, but I had both types locally so thought I should support both.

Review by commit

##### RISK ASSESSMENT / QA PLAN

I don't believe this is high risk.

### CODE REVIEW

File: corehq/apps/userreports/reports/filters/choice_providers.py [hunk 1]
```diff
@@ -186,6 +186,13 @@ def get_values_for_query(self, query_context):
     def get_choices_for_known_values(self, values, user):
         return []
 
+    def _make_choice_from_value(self, value):
+        if value is None:
+            return Choice(NONE_CHOICE, '[Missing]')
+        elif not value:
+            return Choice(EMPTY_CHOICE, '[Empty]')
```
Comment (czue) [inline corehq/apps/userreports/reports/filters/choice_providers.py:193]: If these values aren't sent then `select2` appears to just throw them away, which is why I couldn't just use an empty string here.

File: corehq/apps/userreports/reports/filters/choice_providers.py [hunk 2]
```diff
@@ -186,6 +186,13 @@ def get_values_for_query(self, query_context):
     def get_choices_for_known_values(self, values, user):
         return []
 
+    def _make_choice_from_value(self, value):
+        if value is None:
+            return Choice(NONE_CHOICE, '[Missing]')
+        elif not value:
```
Comment (calellowitz) [inline corehq/apps/userreports/reports/filters/choice_providers.py:192]: Does this cause issues with other falsy values like 0 or false, which could be in UCRs

**Review by calellowitz — APPROVED** [review]
Left a question but if thats fine, then this looks good

**Review by calellowitz — APPROVED** [review]

**Review by esoergel — APPROVED** [review]
Did you consider combining "missing" and "empty" into one filter?  I'm not sure users will know what the difference is, and might prefer a single option that encompasses both.  For example, you can't easily delete a case property, so users often set it to `""`, which I think they consider to be the same thing.

**Comment by czue** [comment]
> Did you consider combining "missing" and "empty" into one filter? I'm not sure users will know what the difference is, and might prefer a single option that encompasses both. For example, you can't easily delete a case property, so users often set it to "", which I think they consider to be the same thing.

Yeah I think this is better UX and if I had more time I'd likely have tried to do that (and I think it should be possible now that I have my head wrapped around the different parts a bit better). Agree the distinction is unlikely to matter to ~anyone and so merging them makes sense.

I'll make a note to look into updating it when I have time.

**Comment by czue** [comment]
@esoergel https://github.com/dimagi/commcare-hq/pull/27710

**Pull request merged**